File: Top-Interview-Questions/DynamicProgramming-Easy-Collection/53-maximum-subarray.cpp
// TC: O(n)
// SC: O()
// Runtime: 12 ms, faster than 64.48%
// Memory Usage: 13.4 MB, less than 22.87%
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        
        if(!nums.size())
            return 0;
        
        int max = INT_MIN, sum = 0;
        
        for(int i = 0; i < nums.size(); i++){
            sum += nums[i];
            
            if(sum > max)
                max = sum;
            
            if(sum < 0)
                sum = 0;
        }
        
        return max;
    }
};

File: Top-Interview-Questions/DynamicProgramming-Easy-Collection/53-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        curr_sum = nums[0]
        max_sum = nums[0]

        for i in range(1, n):
            curr_sum = max(nums[i], curr_sum + nums[i])
            max_sum = max(max_sum, curr_sum)
            
        return max_sum   
    # TC: O(N)
    # SC: O(1)
    # Runtime: 68 ms, faster than 67.62%
    # Memory Usage: 14.6 MB, less than 43.17%

    '''
        n = len(nums)
        max_sum = nums[0]
        for i in range(1, n):
            if nums[i - 1] > 0:
                nums[i] += nums[i - 1] 
            max_sum = max(nums[i], max_sum)

        return max_sum
    '''